src/bhfmm/containers/UniformPseudoParticleContainer.cpp

```diff
@@ -3483,13 +3478,13 @@ void UniformPseudoParticleContainer::AllReduceLocalMoments(int mpCells, int _cur
 
 	}
 
-	std::fill(_coeffVector, _coeffVector + _coeffVectorLength * 2, 0.0);
+	std::fill(_coeffVector.begin(), _coeffVector.end(), 0.0);
 
 #endif
 	global_simulation->timers()->stop("UNIFORM_PSEUDO_PARTICLE_CONTAINER_ALL_REDUCE_ME");
 }
 
-void UniformPseudoParticleContainer::getHaloValues(Vector3<int> localMpCellsBottom,int bottomLevel, double *buffer,
+void UniformPseudoParticleContainer::getHaloValues(Vector3<int> localMpCellsBottom,int bottomLevel, std::vector<double> buffer,
```

Comment:
Here I'm VERY confused: `localMpCellsBottom` is passed as a copy and four lines further down it is copied again into `localMpCells`, which is a vector that only exists in this function. This is then updated.
1. The initial input surely has to be a `const &` since we are copying it anyway.
2. Does it make sense to copy it again!? Don't we want the update to the local expansions? No idea if this is a bug or intended behavior...

Either way, at least these changes should be done to avoid unnecessary copies:
```suggestion
void UniformPseudoParticleContainer::getHaloValues(const Vector3<int> &localMpCellsBottom,int bottomLevel, std::vector<double> &buffer,
```
(I suppose we can't make `buffer` a `const` parameter because the functions it is passed to don't take `buffer` as `const` even though they probably don`t change it.)

Reply:
I have no idea regarding your second point. Indeed, it looks a bit weird. But I wouldn't change it for now, as I do not totally understand the whole class and its methods' interactions.
I think this method's main task is to write to `buffer`. Maybe the variable is copied to `localMpCells` since it is changed during the `for` loop but the "external" variable `localMpCellsBottom` should not change (wouldn't anyway due to the copy).

https://github.com/ls1mardyn/ls1-mardyn/blob/f347b7fb52a25fbcf544356947ceac87df736519/src/bhfmm/containers/UniformPseudoParticleContainer.h#L192-L203

Reply:
> I think this method's main task is to write to buffer. Maybe the variable is copied to localMpCells since it is changed during the for loop but the "external" variable localMpCellsBottom should not change (wouldn't anyway due to the copy).

This is also my guess. Then let's only change the passed parameter to a `const &` to indicate that what is passed into the function does not change and to avoid one copy but keep the copy in the function.